raiden/utils/upgrades.py

```diff
@@ -0,0 +1,127 @@
+import os
+import shutil
+import sqlite3
+from pathlib import Path
+
+import filelock
+import sqlitebck
+import structlog
+
+from raiden.exceptions import RaidenDBUpgradeError
+from raiden.storage.serialize import JSONSerializer
+from raiden.storage.sqlite import RAIDEN_DB_VERSION, SQLiteStorage
+from raiden.storage.versions import older_db_file
+from raiden.utils.migrations.v16_to_v17 import upgrade_initiator_manager
+
+UPGRADES_LIST = [
+    upgrade_initiator_manager,
+]
+
+
+log = structlog.get_logger(__name__)
+
+
+def get_file_lock(db_filename: Path):
+    lock_file_name = f'{db_filename}.lock'
+    return filelock.FileLock(lock_file_name)
+
+
+def get_db_version(db_filename: Path):
+    db_lock = get_file_lock(db_filename)
+    with db_lock:
+        # Perform a query directly through SQL rather than using
+        # storage.get_version()
+        # as get_version will return the latest version if it doesn't
+        # find a record in the database.
+        conn = sqlite3.connect(
+            str(db_filename),
+            detect_types=sqlite3.PARSE_DECLTYPES,
+        )
+        cursor = conn.cursor()
+        try:
+            cursor.execute('SELECT value FROM settings WHERE name=?;', ('version',))
+            query = cursor.fetchall()
+            if len(query) == 0:
+                return 0
+            return int(query[0][0])
+        except sqlite3.OperationalError:
+            return 0
+
+
+class UpgradeManager:
+    """ This class is responsible for figuring out which migrations
+    need to be executed in order to bring the database up to date
+    with the current implementation.
+    """
+    def __init__(self, db_filename: str):
+        self._current_db_filename = Path(db_filename)
+
+    def run(self):
+        """
+        The `_current_db_filename` is going to hold the filename of the database
+        with the new version. However, the previous version's data
+        is going to exist in a file whose name contains the old version.
+        Therefore, running the migration means that we have to copy
+        all data to the current version's database, execute the migration
+        functions.
+        """
+        if get_db_version(self._current_db_filename) == RAIDEN_DB_VERSION:
+            # The current version has already been created / updraded.
+            return
+        else:
+            # The version inside the current database was not the expected one.
+            # Delete and re-run migration
+            self._delete_current_db()
```

Comment:
Hummm, this does not look right, every upgrade expect the version in the database to be the wrong one, otherwise an upgrade wouldn't be necessary.

The current database is the used by Raiden in a previous run, correct? If the existing database is deleted before every upgrade, the net effect is that a new empty database is created.

Reply:
Assuming the current version is 17. If we get the version from this file and it turns out to be 16 (which shouldn't happen but if it does, it's wrong) then we delete the file and migrate to 17.

So the "current DB"'s version should always match `RAIDEN_DB_VERSION`